### DISCUSSION TITLE
Deployer 7, rsync, cwd

### DISCUSSION THREAD

**Discussion by teal-bauer** [deployphp/deployer#3604] [Help needed] [answered]
I am running deployer 7 within a docker container. The container has the workdir set to /var/www/html/htdocs, but the Git repository root is /var/www/html. I cannot find a way to run rsync relative to the Git repository root - I tried combinations of `cd`, `chdir`, `runLocally(...)`, to no avail. Also setting `rsync_src` does not seem to work.

This is especially painful in the `push` recipe, which just won't work because git returns filenames like 'htdocs/index.php' but rsync looks for that relative to 'htdocs/' and fails.

**Comment by antonmedv** [comment]
I didn't get the question. `runLocally('rsync path/ to/');`

**Comment by teal-bauer** [comment]
Well, I was making a recipe similar to the packaged `push` recipe which calls `upload()`, so I cannot influence the rsync call directly (and I don't really want to duplicate more code). A plain `upload()` should have the same issue though - no way to set its working directory. I can monkey patch the DEPLOYER_ROOT but that seems wrong, too.

**Reply by antonmedv** [reply]
I not fully understand the problem. Maybe just reconfigure Apache?

**Comment by teal-bauer** [comment]
Hmm ok, let me try again :)

We have a docker container with this minimal Dockerfile:

```
FROM php:8.0.28
RUN set -ex; \
    wget -qO /usr/local/bin/deployer https://github.com/deployphp/deployer/releases/download/v7.3.1/deployer.phar; \
    chmod +x /usr/local/bin/deployer
VOLUME [ "/var/www/html" ]
WORKDIR /var/www/html/htdocs/
```

Mounted into the `/var/www/html` volume is a Git repository with the following structure:

```
.
htdocs/deploy.php
var/
...other files...
```

I run `deployer push staging` within the container:

```
docker run --rm -it our-image deployer push staging
```

Deployer starts with CWD `/var/www/html/htdocs`, where `deploy.php` also lies. But the Git root is one directory up from that.

I defined a task `push:all` like this:

```php
task('push:all', static function () {
  $local_base = trim(runLocally("git rev-parse --show-toplevel"));
  $files = explode("\n", runLocally("git ls-files --others --modified --exclude-standard", ['cwd' => $local_base]));

  if (count($files) == 0 || empty($files[0])) {
    throw new Deployer\Exception\GracefulShutdownException('No new or modified files in work directory, nothing to do');
  }

  info('uploading:');
  foreach ($files as $file) {
    writeln(" - $file");
  }

  upload(
    $files,
    '{{current_path}}',
    ['progress_bar' => false, 'options' => ['--relative']]
  );

  // Mark this release as dirty.
  run("echo '{{user}}' > {{current_path}}/DIRTY_RELEASE");
  info('done');
});
```

Basically copying the push function that ships with deployer, with two differences: 1) it also catches git-untracked files, and 2) it always examines the git repository root instead of starting from `DEPLOYER_ROOT`.

Here's the issue:

* DEPLOYER_ROOT and CWD point to the `htdocs/` subdirectory of the repo
* There is no way that I found to tell `upload(...)` to use a different working directory
* So the upload() call fails because git ls-files returns repo-relative names (e.g. `htdocs/deploy.php`), but `upload(...)` expects paths relative to `DEPLOYER_ROOT`

There are some workarounds, but they all feel icky, namely:

1. Override workdir in docker call (option `-w`): Works, but user has to remember it, annoying typing
2. Change WORKDIR in Dockerfile: Works, but annoying for all other dev work which is almost always concerned with files in `htdocs/`
3. Construct my own rsync call: Works, but messy and annoying, would have to duplicate even more code from Deployer to get it right
4. Clobber `DEPLOYER_ROOT` from within the task (getenv/putenv): Works but ... well. It feels like manipulating internal state.
5. Override DEPLOYER_ROOT externally (e.g. by passing an env to Docker): See 1)

I would hope/expect that upload would take a `cwd` option or there is an explicit way to set the working_dir otherwise. The CLI documentation seems to imply that all commands could take a cwd arg, but in reality it's only runLocally. within() also doesn't work for upload(). Alternatively, having a function to explicitly set/override the local working dir would be ok too.

**Reply by antonmedv** [reply]
It seems like you have quite a detailed understanding of the issue at hand and have tried a number of different potential solutions. Unfortunately, the `upload()` function in Deployer doesn't support changing the working directory directly.

Although it is not ideal, the approach of manipulating `DEPLOYER_ROOT` (your option 4) might be the most straightforward in this case if you want to stick with using Deployer's built-in functions. Despite the fact that it feels like manipulating internal state, it would provide the pathing that the `upload()` function needs to operate correctly.

Here's how you could manipulate `DEPLOYER_ROOT` from within the task:

```php
task('push:all', static function () {
  $local_base = trim(runLocally("git rev-parse --show-toplevel"));
  $files = explode("\n", runLocally("git ls-files --others --modified --exclude-standard", ['cwd' => $local_base]));

  if (count($files) == 0 || empty($files[0])) {
    throw new Deployer\Exception\GracefulShutdownException('No new or modified files in work directory, nothing to do');
  }

  // Change DEPLOYER_ROOT
  set('deploy_path', $local_base);

  info('uploading:');
  foreach ($files as $file) {
    writeln(" - $file");
  }

  upload(
    $files,
    '{{current_path}}',
    ['progress_bar' => false, 'options' => ['--relative']]
  );

  // Mark this release as dirty.
  run("echo '{{user}}' > {{current_path}}/DIRTY_RELEASE");
  info('done');
});
```

I understand your frustration, however. This seems like a missing feature in Deployer and you might want to consider submitting an issue or even a pull request to the Deployer project to support a `cwd` option in the `upload()` function, if such a feature hasn't been implemented already after my knowledge cutoff.

**Reply by teal-bauer** [reply, marked as answer]
I don't know your knowledge cutoff ^^ but the code ChatGPT generated is wrong, because `deploy_path` refers to the target system. Like I said, what needs to be done is a song-and-dance with putenv, like so:

```php
  $old_root = getenv('DEPLOYER_ROOT') ?: __DIR__;
  putenv("DEPLOYER_ROOT={$local_base}");
  upload(
    $files,
    '{{current_path}}',
    ['progress_bar' => false, 'options' => ['--relative']]
  );
  putenv("DEPLOYER_ROOT={$old_root}");
```

While I don't mind using ChatGPT to formulate responses, I think removing the obvious giveaways like "my knowledge cutoff" and maybe sanity checking the answer might be useful... ;)

**Reply by antonmedv** [reply]
Nice 👌🏻 well, chatgpt answers right up to 90%))
